### PR TITLE
Change variable map to repeated map entries for ordering

### PR DESCRIPTION
**Pull request by mayitbeegh**
Signed-off-by: Sean Lin <sean@union.ai>

# TL;DR
Turn variable map into repeated map entries for ordering while maintaining backwards compatibility See https://developers.google.com/protocol-buffers/docs/proto3#backwards_compatibility
Related PR:
- flytectl: https://github.com/flyteorg/flytectl/pull/158
- flyteplugin: https://github.com/flyteorg/flyteplugins/pull/198
- flytepropeller: https://github.com/flyteorg/flytepropeller/pull/305
- flyteadmin: https://github.com/flyteorg/flyteadmin/pull/235
- flytekit: https://github.com/flyteorg/flytekit/pull/608

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/940

## Follow-up issue
_NA_

### CODE REVIEW

File: protos/flyteidl/core/interface.proto [hunk 1]
```diff
@@ -16,10 +16,16 @@ message Variable {
     string description = 2;
 }
 
-// A map of Variables
+// A map of Variables in the form of repeated map entries
 message VariableMap {
     // Defines a map of variable names to variables.
-    map<string, Variable> variables = 1;
+    repeated VariableMapFieldEntry variables = 1;
```
Comment (kumare3) [inline protos/flyteidl/core/interface.proto:22]: isnt this a breaking change?

Comment (kumare3) [reply protos/flyteidl/core/interface.proto:22]: if not, so we realized that variables and parameters being separate was not a good idea, we should just have one - one with default values

File: protos/flyteidl/core/interface.proto [hunk 2]
```diff
@@ -16,10 +16,16 @@ message Variable {
...
+}
+
+// Map entry of VariableMap
+message VariableMapFieldEntry {
+    string key = 1;
```
Comment (EngHabu) [inline protos/flyteidl/core/interface.proto:27]: I don't think the field name matter, does it? if so, can we call it `name`?

File: protos/flyteidl/core/interface.proto [hunk 3]
```diff
@@ -16,10 +16,16 @@ message Variable {
...
+    Variable value = 2;
```
Comment (EngHabu) [inline protos/flyteidl/core/interface.proto:28]:
```suggestion
    Variable var = 2;
```

File: protos/flyteidl/core/interface.proto [hunk 4]
```diff
@@ -16,10 +16,16 @@ message Variable {
...
-// A map of Variables
+// A map of Variables in the form of repeated map entries
 message VariableMap {
     // Defines a map of variable names to variables.
-    map<string, Variable> variables = 1;
+    repeated VariableMapFieldEntry variables = 1;
+}
+
+// Map entry of VariableMap
+message VariableMapFieldEntry {
```
Comment (EngHabu) [inline protos/flyteidl/core/interface.proto:26]:
```suggestion
message VariableMapEntry {
```

File: protos/flyteidl/core/interface.proto [hunk 5]
```diff
@@ -44,8 +50,14 @@ message Parameter {
...
 message ParameterMap {
     // Defines a map of parameter names to parameters.
-    map<string, Parameter> parameters = 1;
+    repeated ParameterMapFieldEntry parameters = 1;
+}
+
+// Map entry of ParameterMap
+message ParameterMapFieldEntry {
+    string key = 1;
+    Parameter value = 2;
```
Comment (EngHabu) [inline protos/flyteidl/core/interface.proto:62]: ditto

**Review by EngHabu — COMMENTED** [review]
Awesome! While this isn't a wire-format breaking change, it's a generated-code breaking change... let's not merge it until we have all other PRs ready to go...

File: protos/flyteidl/core/interface.proto (hunk 5 above)
Comment (katrogan) [reply protos/flyteidl/core/interface.proto]: @EngHabu did you want this to be "var"? seems kinda incongruous with `Parameter`

**Review by kumare3 — APPROVED** [review]
